benchmarks/simple_rma.cpp

```diff
@@ -0,0 +1,342 @@
+                    fields.back().push_back(
+                        gridtools::ghex::wrap_field<gridtools::ghex::cpu,2,1,0>(
+                            local_domains.back(),
+                            fields_raw.back().back().data(),
+                            offset,
+                            local_ext_buffer));
+#ifdef __CUDACC__
+                    fields_raw_gpu.back().push_back( 
+                        std::unique_ptr<T,cuda_deleter<T>>{
+                            [this](){ void* ptr; cudaMalloc(&ptr, max_memory*sizeof(T)); return (T*)ptr; }()});
+                    fields_gpu.back().push_back(
+                        gridtools::ghex::wrap_field<gridtools::ghex::gpu,2,1,0>(
+                            local_domains.back(),
+                            fields_raw_gpu.back().back().get(),
+                            offset,
+                            local_ext_buffer));
+#endif
+                }
+                comms.push_back(context.get_communicator(context.get_token()));
+            }
+        }
+
+        pattern = std::unique_ptr<pattern_type>{
+            new pattern_type{
+                gridtools::ghex::make_pattern<gridtools::ghex::structured::grid>(
+                        context, halo_gen, local_domains)}};
+        
+        cos.resize(num_threads);
+#pragma omp parallel
+        {
+            const auto j = omp_get_thread_num();
+
+            auto bco = gridtools::ghex::bulk_communication_object<
+                gridtools::ghex::structured::rma_range_generator,
+                pattern_type,
+#ifndef __CUDACC__
+                field_type
+#else
+                gpu_field_type
+#endif
+            > (comms[j]);
+
+#ifndef __CUDACC__
+            for (int i=0; i<num_fields; ++i)
+                bco.add_field(pattern->operator()(fields[j][i]));
+#else
+            for (int i=0; i<num_fields; ++i)
+                bco.add_field(pattern->operator()(fields_gpu[j][i]));
+#endif
+            
+            cos[j] = std::move(bco);
+        }
+    }
+
+    void exchange()
+    {
+#pragma omp parallel
+        {
+
+            const auto j = omp_get_thread_num();
+                std::this_thread::sleep_for(std::chrono::milliseconds(50));
+                {
+                    std::lock_guard<std::mutex> io_lock(io_mutex);
+                    std::cout << "Thread #" << j << ": on CPU " << sched_getcpu() << "\n";
+                }
+    
+                // warm up
+                for (int t = 0; t < 50; ++t)
+                {
+                    cos[j].exchange().wait();
+                }
+    
+                auto start = clock_type::now();
+                for (int t = 0; t < num_reps; ++t)
+                {
+                    cos[j].exchange().wait();
+                }
+                auto end = clock_type::now();
+                std::chrono::duration<double> elapsed_seconds = end - start;
+                
+                if (comm.rank() == 0 && j == 0)
+                {
```

Comment:
Agreed, I'll look into that.

Reply:
This one was still pending...

Reply:
yes, working on that atm.

Reply:
I will leave this as is for now, because the communication is synchronized anyways, so it does not matter too much if I print the other threads/ranks imo.